index.rst

```diff
@@ -302,107 +305,152 @@ Is this case, we could have something like:
-The configuration schema contains default values for the configuration which are loaded if the ``start`` command is sent with an empty ``configuration`` attribute (the default value).
+The configuration schema currently contains default values for the configuration parameters which are loaded if the ``start`` command is sent with an empty ``configuration`` attribute (the default value).
 Nevertheless, the values in the schema are seldom valid beyond a unit testing environment, which requires users to provide some kind of *operational defaults* or *default label*.
-One can see how this can cause confusion when operating the system since now "default" has two different meanings, e.g.; *schema default* and *operational default*.
+One can see how this can cause confusion when operating the system since "default" can be interpreted in two different ways, e.g.; *schema default* and *operational default*.
 Furthermore, it is usually enough to override a small subset of the *schema defaults* for operations.
 Therefore, to get a full set of applied configurations, users must look at two distinct repositories; the configuration repository (for the modified parameters) and the CSC repository (for the schema defaults).
 
 The proposal to improve this aspect of the system is:
 
-#.  Remove default values from the configuration schema.
+#.  Remove all default values from configuration schema definition in the CSC repository.
 
     - See this :download:`example schema <_static/ATSpectrograph_schema.yaml>` for the ATSpectrograph CSC.
+    - Unit tests will need to utilize configuration files stored in the `tests/data/config` directory as is done for the `ATDome CSC <https://github.com/lsst-ts/ts_ATDome/tree/develop/tests/data/config>`_. See `Salobj documentation <https://ts-salobj.lsst.io>`__) for more details.
+
+#.  In the configuration repository for the given CSC (e.g `ts_config_attcs <https://github.com/lsst-ts/ts_config_attcs>`_ for the ATDome) there shall be a ``_init.yaml`` file defining all values that are expected to be relatively static in operations (we intentionally use "_init" instead of "_default").
+
+    - See this :download:`example _init.yaml <_static/_init.yaml>` for the ATSpectrograph CSC.
+    - This file is the first configuration file to get loaded by the CSC
+    - It is not possible to load this file using the `configurationToApply` attribute
+    - Note that all CSCs having multiple algorithms, each with different required attributes, must have an initial set of defaults in this file.
+
+#.  Also in the configuration repository for the given CSC, there shall be a file corresponding to each site where the CSC is used (e.g. ``_summit.yaml, _ncsa.yaml, _base.yaml``).
+    These files contain site specific configuration parameters such as IP addresses and ports.
+    SalObj determines what site file should be loaded automatically by parsing the ``LSST_DDS_PARTITION_PREFIX`` environment variable
```

Comment:
We still need to decide what to do about unit tests, which set LSST_DDS_PARTITION_PREFIX to a long random value.

Also: do we really want to require site-specific files when they are not needed? I can see arguments both ways.

Some options:
* Only load `_<site>.yaml` if there is one that matches LSST_DDS_PARTITION_PREFIX. That is dirt simple, and allows us to omit `_<site>.yaml` files if not needed. We would also probably want an official set of site names so we can test that all `_<site>.yaml` files have one of the blessed names (i.e. we don't want `_sumit.yaml` to be silently ignored).
* Require that all site names meet some simple criterion, such as "only (ASCII?) letters and digits, starting with a letter" or "a valid python identifier that does not start with _". If LSST_DDS_PARTITION_PREFIX meets that criteria then it is treated as a site name and the `_<site>.yaml` file must exist; otherwise salobj won't even look for such a file.

Reply:
I don't think we need to require a `_<site>.yaml,` but then obvious `_init.yaml` needs to be complete.

I prefer option 1. The issue is we'll need to create some sort of custom "site" that we can use for if we want to stand something up temporarily. Essentially have a list of blessed names, where it includes a `tmp*` wildcard or something.

Reply:
I prefer option 1 as well... I think we could add "`citest`" and/or "`test`" to the valid cite options (for the occasional test).

Reply:
Instead os starting the phrase with "SalObj", maybe start with "The CSC"? just want to make sure we leave it open to implementing this in other cases as well.

Reply:
OK. I also prefer 1. Should we maintain a blessed list of site names, so we can catch typos in site file names (e.g. `_sumit.yaml`) and if so, where? I am willing to stuff it into ts_salobj since the config unit tests rely on that package. I don't think ts_xml, ts_idl or ts_ddsconfig are obvious places.

Reply:
I'm reluctant to include yaml files in salobj (having just gotten rid of the yaml schema files) because it's a nuisance to deploy the software. I'd rather make it code (my first choice) or include that data in some data-centric package such as ts_xml or ts_idl. The trouble is that none of our data-centric packages seem appropriate.